### PR TITLE
DT-1561 Add cron job for queue housekeeping and associated endpoints

### PR DESCRIPTION
**Pull request by karenmillermoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/resource/QueueResource.kt [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  @ApiResponses(
+    value = [
+      ApiResponse(responseCode = "401", description = "Unauthorised, requires a valid Oauth2 token"),
+      ApiResponse(responseCode = "403", description = "Forbidden, requires an authorisation with role PTPU_QUEUE_ADMIN")
+    ]
+  )
+  fun transferEventDlq(): Unit = queueAdminService.transferEventMessages()
+
+  @PutMapping("/purge-event-dlq")
+  @PreAuthorize("hasRole('PTPU_QUEUE_ADMIN')")
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/resource/QueueResource.kt:33]: Reminder that we'll need some tests around authorisation...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/resource/QueueResource.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      ApiResponse(responseCode = "401", description = "Unauthorised, requires a valid Oauth2 token"),
+      ApiResponse(responseCode = "403", description = "Forbidden, requires an authorisation with role PTPU_QUEUE_ADMIN")
+    ]
+  )
+  fun purgeEventDlq(): Unit = queueAdminService.clearAllDlqMessagesForEvent()
+
+  @PutMapping("/queue-housekeeping")
+  @Operation(
+    summary = "Triggers maintenance of the queues",
+    description = "This is an internal service which isn't exposed to the outside world. It is called from a Kubernetes CronJob named `index-housekeeping-cronjob`"
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/resource/QueueResource.kt:49]: Small typo - we renamed the cronjob but forgot to update this comment

**Review by andymarke — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/resource/QueueResource.kt (hunk 2 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/resource/QueueResource.kt:49]: good spot

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/resource/QueueResource.kt (hunk 1 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/resource/QueueResource.kt]: added

File: src/testIntegration/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/QueueAdminServiceTest.kt [hunk 3]
```diff
@@ -90,40 +90,55 @@ internal class QueueAdminServiceTest {
...
     }
 
     private fun stubDlqMessageCount(count: Int) =
       whenever(eventAwsSqsDlqClient.getQueueAttributes(eventDlqUrl, listOf("ApproximateNumberOfMessages")))
         .thenReturn(GetQueueAttributesResult().withAttributes(mutableMapOf("ApproximateNumberOfMessages" to count.toString())))
   }
 
-  fun offenderChangedMessage(crn: String) =
+  @Nested
+  inner class ClearAllDlqMessagesForEvent {
```
[the remark this answers is not in the document]
Comment (karenmillermoj) [reply src/testIntegration/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/services/QueueAdminServiceTest.kt:121]: added

File: helm_deploy/prison-to-probation-update/templates/ingress.yaml [hunk 4]
```diff
@@ -17,6 +17,10 @@ metadata:
         deny all;
         return 401;
       }
+      location /queue_admin/queue-housekeeping {
```
Comment (mikehalmamoj) [inline helm_deploy/prison-to-probation-update/templates/ingress.yaml:20]: typo?

File: src/testIntegration/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/MessageIntegrationTest.kt [hunk 5]
```diff
@@ -20,13 +24,22 @@ import javax.inject.Inject
     "prisontoprobation.message-processor.enabled=true",
     "prisontoprobation.message-processor.delay=50",
     "prisontoprobation.hold-back.duration=0m"
-  ]
+  ],
+  webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT
```
Comment (mikehalmamoj) [inline src/testIntegration/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/MessageIntegrationTest.kt:28]: As you've moved the tests I think you don't need the web server starting, or the WebTestClient

Comment (mikehalmamoj) [reply src/testIntegration/kotlin/uk/gov/justice/digital/hmpps/prisontoprobation/MessageIntegrationTest.kt:28]: Nope that's not true is it! :facepalm:  Please ignore.

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**